src/rust/inetstack/protocols/tcp/established/ctrlblk.rs

```diff
@@ -115,44 +112,40 @@ struct Receiver {
     pub receive_next: SeqNumber,
 
     // Receive queue.  Contains in-order received (and acknowledged) data ready for the application to read.
-    recv_queue: VecDeque<DemiBuffer>,
+    recv_queue: AsyncQueue<DemiBuffer>,
 }
 
 impl Receiver {
     pub fn new(reader_next: SeqNumber, receive_next: SeqNumber) -> Self {
         Self {
             reader_next,
             receive_next,
-            recv_queue: VecDeque::with_capacity(RECV_QUEUE_SZ),
+            recv_queue: AsyncQueue::with_capacity(RECV_QUEUE_SZ),
         }
     }
 
-    pub fn pop(&mut self, size: Option<usize>) -> Result<Option<DemiBuffer>, Fail> {
-        // Check if the receive queue is empty.
-        if self.recv_queue.is_empty() {
-            return Ok(None);
-        }
-
+    pub async fn pop(&mut self, size: Option<usize>, yielder: Yielder) -> Result<DemiBuffer, Fail> {
         let buf: DemiBuffer = if let Some(size) = size {
-            let buf: &mut DemiBuffer = self.recv_queue.front_mut().expect("receive queue cannot be empty");
+            let mut buf: DemiBuffer = self.recv_queue.pop(yielder).await?;
             // Split the buffer if it's too big.
             if buf.len() > size {
                 buf.split_front(size)?
             } else {
-                self.recv_queue.pop_front().expect("receive queue cannot be empty")
+                // I think this was a bug before.
```

Comment:
I agree, but I don't we just add a debug-level message here? If we notice something bad we keep track of this.